// stylistic/lib/functions.js
/* Argument Types */

'use strict';

exports.__esModule = true;
function px(n) {
  if (/%$/.test(n)) return n;
  return (n || 0) + 'px';
}

function ratio(n) {
  return n;
}

function string(s) {
  return s;
}

function colorComponent(x) {
  return x < 0 ? 0 : x > 255 ? 255 : Math.round(x);
}

function alpha() {
  var x = arguments.length <= 0 || arguments[0] === undefined ? 1 : arguments[0];

  return x < 0 ? 0 : x > 1 ? 1 : x;
}

function deg(n) {
  return (n || 0) + 'deg';
}

function cssFunction(name) {
  for (var _len = arguments.length, params = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    params[_key - 1] = arguments[_key];
  }

  return function () {
    for (var _len2 = arguments.length, values = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      values[_key2] = arguments[_key2];
    }

    return name + '(' + params.map(function (p, i) {
      return p(values[i]);
    }) + ')';
  };
}

/* Functions */

var url = cssFunction('url', string);
exports.url = url;
var rgb = cssFunction('rgb', colorComponent, colorComponent, colorComponent);
exports.rgb = rgb;
var rgba = cssFunction('rgba', colorComponent, colorComponent, colorComponent, alpha);
exports.rgba = rgba;
var translate = cssFunction('translate', px, px);
exports.translate = translate;
var translate3d = cssFunction('translate3d', px, px, px);
exports.translate3d = translate3d;
var translateX = cssFunction('translateX', px);
exports.translateX = translateX;
var translateY = cssFunction('translateY', px);
exports.translateY = translateY;
var translateZ = cssFunction('translateZ', px);
exports.translateZ = translateZ;
var scale = cssFunction('scale', ratio, ratio);
exports.scale = scale;
var scale3d = cssFunction('scale3d', ratio, ratio, ratio);
exports.scale3d = scale3d;
var scaleX = cssFunction('scaleX', ratio);
exports.scaleX = scaleX;
var scaleY = cssFunction('scaleY', ratio);
exports.scaleY = scaleY;
var scaleZ = cssFunction('scaleZ', ratio);
exports.scaleZ = scaleZ;
var rotate = cssFunction('rotate', deg);
exports.rotate = rotate;
var rotate3d = cssFunction('rotate3d', ratio, ratio, ratio, deg);
exports.rotate3d = rotate3d;
var rotateX = cssFunction('rotateX', deg);
exports.rotateX = rotateX;
var rotateY = cssFunction('rotateY', deg);
exports.rotateY = rotateY;
var rotateZ = cssFunction('rotateZ', deg);
exports.rotateZ = rotateZ;
var skew = cssFunction('skew', deg, deg);
exports.skew = skew;
var skewX = cssFunction('skewX', deg);
exports.skewX = skewX;
var skewY = cssFunction('skewY', deg);
exports.skewY = skewY;
